atlas_metadata_validator/atlas_checker.py

```diff
@@ -204,10 +229,31 @@ def run_singlecell_checks(self, logger):
         for protocol in sc_protocol_values:
             if protocol.lower() in library_construction_terms.get("droplet"):
                 droplet_terms = get_controlled_vocabulary("required_droplet_sdrf_fields", "atlas", logger)
+                allowed_read_values = get_controlled_vocabulary("allowed_read_values", "atlas", logger)
                 for dt in droplet_terms:
                     if dt not in self.sdrf_values:
                         logger.error("Required SDRF droplet field \"{}\" not found.".format(dt))
                         self.errors.add("SC-E10")
+                    else:
+                        # Check that the values match the allowed
+                        droplet_term_values = {row[i] for row in self.sdrf for i, c in enumerate(self.sdrf_header)
+                                               if self.normalise_header(dt) == self.normalise_header(c)}
+                        if dt.endswith("read"):
+                            for droplet_value in droplet_term_values:
+                                if droplet_value not in allowed_read_values:
+                                    logger.error(f"Read value \"{droplet_value}\" for \"{dt}\" is not allowed.")
+                                    self.errors.add("SC-E13")
```

Comment:
Again, this may be a little convoluted. Could you apply some python set difference magic instead (think it's just .difference() ), between observed and allowed values, rather than all the looping?

Reply:
Good thinking! I applied the set comparison and moved duplicated code to a function. Though I only managed to get rid of one loop. I need the outer one to keep track of the field name and the inner one to loop through the error logging (to get one line per wrong value).